Support New Mape Heat Branding

brab opened the pull request:
Support new thermostats under the Mape Heat brand. The library defaults to using the original NuHeat API if no new arguments are used.

closes #4

brab committed:
Extend config to support MapeHeat

ref #4

NuHeat is now branded as MapeHeat and new thermostats are sold under the
new brand, however old thermostats appear to continue to be supported
with their original NuHeat brand. There is a new URL that uses the same
API for MapeHeat products and so this library should support both, with
NuHeat remaining as the default.

brab committed:
Update tests

Adjusting the tests made it clear that new utility functions would be
helpful in reducing repetition for getting URLs and request headers.

brab commented:
@broox let me know what you think or if you'd like any changes made

broox commented on nuheat/config.py:
```diff
@@ -1,15 +1,12 @@
-    "Host": "mynuheat.com",
-    "DNT": "1",
-    "Origin": "https://mynuheat.com/api"
+BRANDS = ("NUHEAT", "MAPEHEAT")
```
it'd be nice if we defined these individual strings as constants as well so that we can use the constants (at least the default, "NUHEAT" one) in code, rather than strings.

broox commented on nuheat/api.py:
```diff
@@ -9,17 +9,19 @@
 class NuHeat(object):
 
-    def __init__(self, username, password, session_id=None):
+    def __init__(self, username, password, session_id=None, brand="NUHEAT"):
```
it'd be nice of `NUHEAT` was a constant

broox commented on nuheat/util.py:
```diff
@@ -66,3 +66,29 @@ def nuheat_to_celsius(nuheat_temperature):
     fahrenheit = nuheat_to_fahrenheit(nuheat_temperature)
     return fahrenheit_to_celsius(fahrenheit)
+
+def get_api_url(config, brand="NUHEAT"):
```
rather than passing `config` around as arguments, why not simply `import` it in this file?

broox commented:
Thanks for doing this! Just left a couple questions/suggestions for you @brab

brab commented:
Sorry for being unresponsive. I'm hoping to have time this weekend to take a look at your feedback. Thank you for taking a look at it so quickly.

broox commented:
hey @brab - are you still interested in getting this in? if so, please rebase from the latest master. i'm in the process of moving from python2.7+ to python3.7+.

thanks!

brab committed:
Refactor with brand constants

Rather than passing strings around we now have constants to represent
the two brand names

brab committed:
Import config instead of passing as an arg

This style is more consistent with this project

brab commented:
@broox I've rebased, resolved conflicts and addressed your two comments (constants for brand names and importing config). What do you think?

I've also tested manually on my MapeHeat thermostat and it's working well. If you've got a NuHeat one would you mind confirming that it still works for you please?

broox commented on nuheat/api.py:
```diff
@@ -9,17 +9,19 @@
         self.username = username
         self.password = password
         self._session_id = session_id
+        self._brand = brand if brand in config.BRANDS else config.BRANDS[0]
```
🙏

broox commented on nuheat/api.py:
```diff
@@ -38,7 +40,11 @@ def authenticate(self):
         }
-        data = self.request(config.AUTH_URL, method="POST", data=post_data)
+        data = self.request(
+            url=util.get_auth_url(brand=self._brand),
```
if taking my suggestion below, this could just be 

```suggestion
            url=self._auth_url,
```

broox commented on nuheat/api.py:
```diff
@@ -63,7 +69,7 @@ def request(self, url, method="GET", data=None, params=None, retry=True):
         :param retry: Attempt to re-authenticate and retry request if necessary
         """
-        headers = config.REQUEST_HEADERS
+        headers = util.get_request_headers(brand=self._brand)
```
if taking my suggestion below, this could be:

```suggestion
        headers = self._request_headers
```

broox commented on nuheat/config.py:
```diff
@@ -1,15 +1,14 @@
+API_URL = "https://{HOSTNAME}/api"
+
+AUTH_URL = "{API_URL}/authenticate/user"
+THERMOSTAT_URL = "{API_URL}/thermostat"
```
these 3 URLs can be killed if taking my suggestion below.

broox commented on nuheat/thermostat.py:
```diff
@@ -138,7 +139,10 @@ def get_data(self):
         }
-        data = self._session.request(config.THERMOSTAT_URL, params=params)
+        data = self._session.request(
+            url=get_thermostat_url(brand=self._session._brand),
```
with my suggestion below, i believe this could be:

```suggestion
            url=self._url,
```

broox commented on nuheat/thermostat.py:
```diff
@@ -338,4 +342,9 @@ def set_data(self, post_data):
         }
-        self._session.request(config.THERMOSTAT_URL, method="POST", data=post_data, params=params)
+        self._session.request(
+            url=get_thermostat_url(brand=self._session._brand),
```
with my suggestion below, i believe this could be:

```suggestion
            url=self._url,
```

broox commented on nuheat/util.py:
```diff
@@ -68,3 +70,29 @@ def nuheat_to_celsius(nuheat_temperature):
+        "HOST": config.HOSTNAMES[brand],
+        "DNT": "1",
+        "Origin": get_api_url(brand=brand),
+    }
```
just a suggestion... these could be simplified even further if they were properties on the `NuHeat` class.

this would
- save us the trouble of having to check if the `brand` exists in `config.BRANDS` in each method
- simplify the string formatting
- prevent us from needing to include `util` in the class
- simplify URL lookups as they'd all be tied to their respective class instances.

it does pull some `config` strings into the NuHeat class, but i think the simplicity it provides is nice.

e.g.
```
class NuHeat(object):
    ...

    @property
    def _hostname(self):
        return config.HOSTNAMES.get(self.brand)
    
    @property
    def _api_url(self):
        return f"https://{self._hostname}/api"

    @property
    def _auth_url(self):
        return f"{self._api_url}/authenticate/user"

    @property
    def _request_headers(self):
        return {
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "application/json",
            "HOST": self._hostname,
            "DNT": "1",
            "Origin": self._api_url,
        }
```

On the thermostat object, we could add the following...

```
class NuHeatThermostat(object):
    ...
    
    @property
    def _url(self):
        return f"{self._session._api_url}/thermostat"
```

broox commented on tests/test_nuheat.py:
```diff
@@ -81,8 +81,14 @@ def test_get_request(self):
-        self.assertEqual(request_headers["Content-Type"], config.REQUEST_HEADERS["Content-Type"])
+        self.assertEqual(
+            request_headers["Origin"],
+            util.get_request_headers()["Origin"],
```
all header checks could be something like:

```suggestion
            api._request_headers["Origin"],
```

broox commented on tests/test_thermostat.py:
```diff
@@ -126,7 +126,7 @@ def test_get_data(self):
         responses.add(
             responses.GET,
-            config.THERMOSTAT_URL,
+            util.get_thermostat_url(),
```
with the suggested change, the mocks would have to be moved below the initialization of `thermostat`, and then could just be...

```suggestion
            thermostat._url,
```

broox commented on tests/test_thermostat.py:
```diff
@@ -137,7 +137,10 @@ def test_get_data(self):
         }
-        request_url = "{}?{}".format(config.THERMOSTAT_URL, urlencode(params))
+        request_url = "{}?{}".format(
+            util.get_thermostat_url(),
```
and these would obviously all be...

```suggestion
            thermostat._url,
```

brab replied:
I like it. I've got the refactor going, and I agree that it's cleaner, but unfortunately it causes a circular dependency in testing because of how `NuHeatThermostat.__init__()` calls `self.get_data()`. This means that a test response must be created before the thermostat instance is initialized, but the URL isn't available for that until it's initialized. 

What I've done in my next commit is add a helper class method called `NuHeatThermostat.get_url()` (name suggestions welcome, I guess I'm not feeling creative at the moment) to use in tests so that we can get the thermostat URL after initializing the NuHeat API object with its URL but before initializing the NuHeatThermostat one. 

Let me know what you think.

broox replied:
ahhh, interesting. good catch. total oversight of mine.

since it only affects the testing mocks, we can just update the tests to build the string rather than add a helper method to the class simply to serve the tests.

e.g.
```
        responses.add(
            responses.GET,
            config.THERMOSTAT_URL,
            f"{api._api_url}/thermostat",
            status=200,
            body=json.dumps(response_data),
            content_type="application/json"
        )
```

that or make a helper in the testing library...

sorry for the super asynchronous back and forth on this. i may pull your branch and play around a bit.

broox replied:
i pulled your branch, removed that test helper, and added a couple `parameterized` brand tests https://github.com/broox/python-nuheat/pull/14

really appreciate your help on this. let me know what you think!

broox reviewed (commented):
thanks for picking this back up, @brab! i'm excited to be able to add support for Mape.

in looking at the code, i came up with a way to simplify this a bit further. my comments are a little verbose, but that's just because i was going through to ensure that my suggestion would actually work.

i think this change would make everything super clean. thoughts?

brab committed:
Refactor utils as properties

Overall this reduces complexity to help with informal reasoning and
maintenance with only 1 small trade off required for testing
